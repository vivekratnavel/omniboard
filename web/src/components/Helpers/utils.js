
export const capitalize = value => {
  // Capitalize each word
  let split = value.split('_');
  split = split.map(key => {
    if (key.length > 0) {
      return key.charAt(0).toUpperCase() + key.slice(1);
    }

    return '';
  });
  return split.join(' ');
};

export const reorderArray = (array, insertAfter, value) => {
  const insertAfterIndex = array.indexOf(insertAfter);
  const valueIndex = array.indexOf(value);
  if (valueIndex >= 0) {
    array.splice(valueIndex, 1);
    array.splice(insertAfterIndex + 1, 0, value);
  }
};

export const arrayDiff = (arr1, arr2) => arr1.filter(i => !arr2.includes(i));

export const arrayDiffColumns = (arr1, arr2) => arr1.filter(item => !arr2.some(i => i.name === item.name && i.value === item.value));

export const headerText = header => {
  return capitalize(header);
};

export const parseServerError = error => {
  const defaultMessage = 'Error: ';
  let message = '';
  if (error.response) {
    // The request was made and the server responded with a status code
    // that falls out of the range of 2xx
    message = error.response.data.message;
  } else if (error.request) {
    // The request was made but no response was received
    // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
    // http.ClientRequest in node.js
    message = 'No response was received from the server.';
  } else {
    // Something happened in setting up the request that triggered an Error
    message = error.message;
  }

  return defaultMessage + message;
};

export const getFileExtension = fileName => fileName.split('.').splice(-1)[0];

/**
 * Creates a new ArrayBuffer from concatenating two existing ones
 *
 * @param {ArrayBuffer | null} buffer1 The first buffer.
 * @param {ArrayBuffer | null} buffer2 The second buffer.
 * @return {ArrayBuffer | null} The new ArrayBuffer created out of the two.
 */
export const concatArrayBuffers = (buffer1, buffer2) => {
  if (!buffer1) {
    return buffer2;
  }

  if (!buffer2) {
    return buffer1;
  }

  const tmp = new Uint8Array(buffer1.byteLength + buffer2.byteLength);
  tmp.set(new Uint8Array(buffer1), 0);
  tmp.set(new Uint8Array(buffer2), buffer1.byteLength);
  return tmp.buffer;
};

export const resolveObjectPath = (object, path, defaultValue) =>
  path.split('.').reduce((o, p) => o && Object.prototype.hasOwnProperty.call(o, p) ? o[p] : defaultValue, object);

export const getAllPaths = (prefix, data, excludeObjectType = false) => {
  return Object.keys(data).reduce((paths, key) => {
    const newPath = prefix ? `${prefix}.${key}` : key;
    if (!excludeObjectType || typeof data[key] !== 'object') {
      paths = [...paths, newPath];
    }

    if (typeof data[key] === 'object' && !Array.isArray(data[key]) && data[key]) {
      const newPaths = Object.keys(data[key]).reduce((acc, item) => {
        if (typeof data[key][item] !== 'object' && data[key][item]) {
          acc.push(`${newPath}.${item}`);
        }

        return acc;
      }, []);
      const recursivePaths = getAllPaths(newPath, data[key], excludeObjectType);
      return [...paths, ...recursivePaths, ...newPaths];
    }

    return paths;
  }, []);
};

export const getOption = (value, options) => {
  return options.find(option => option.value === value);
};
