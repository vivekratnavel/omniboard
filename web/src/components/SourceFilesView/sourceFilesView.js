import PropTypes from 'prop-types';
import React, {Component} from 'react';
import {Alert, Button} from 'react-bootstrap';
import './sourceFilesView.scss';
import ReactList from 'react-list';
import {
  Accordion,
  AccordionItem,
  AccordionItemTitle,
  AccordionItemBody
} from 'react-accessible-accordion';
import saveAs from 'file-saver';
import backend, {setDbInfo} from '../Backend/backend';
import {ProgressWrapper} from '../Helpers/hoc';
import {concatArrayBuffers, getFileExtension, parseServerError} from '../Helpers/utils';
import {imageExtensions} from '../FilePreview/imageExtensions';
import {FilePreview} from '../FilePreview/filePreview';

const MIN_ACCORDION_HEIGHT = 51;

// File preview limit set to 5MB in bytes
const FILE_PREVIEW_LIMIT = 5242880;

class SourceFilesView extends Component {
  static propTypes = {
    files: PropTypes.array.isRequired,
    type: PropTypes.string.isRequired,
    dbInfo: PropTypes.shape({
      path: PropTypes.string,
      key: PropTypes.string
    }).isRequired,
    runId: PropTypes.number.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      sourceFiles: {},
      isLoadingSourceFiles: false,
      error: '',
      isZipInProgress: false,
      isAccordionDataLoading: false,
      accordionError: ''
    };
  }

  _fetchSourceFiles = () => {
    const {runId, type} = this.props;
    if (runId && ['source_files', 'artifacts'].includes(type)) {
      this.setState({
        isLoadingSourceFiles: true,
        error: ''
      });
      let url = '';
      if (type === 'source_files') {
        url = `api/v1/files/source/${runId}`;
      } else if (type === 'artifacts') {
        url = `api/v1/files/artifacts/${runId}`;
      }

      backend.get(url).then(response => {
        const sourceFiles = response.data.reduce((files, file) => {
          files[file._id] = {
            id: file._id,
            uploadDate: file.uploadDate,
            filepath: file.filename,
            fileLength: file.length,
            chunkSize: file.chunkSize
          };
          return files;
        }, {});
        this.setState({
          sourceFiles,
          isLoadingSourceFiles: false,
          error: ''
        });
      }).catch(error => {
        const message = parseServerError(error);
        this.setState({
          isLoadingSourceFiles: false,
          error: message
        });
      });
    } else {
      this.setState({
        error: 'Invalid run id or type.'
      });
    }
  };

  _downloadFile = (fileId, fileName) => {
    return _event => {
      backend({
        url: `api/v1/files/download/${fileId}/${fileName}`,
        method: 'GET',
        responseType: 'blob'
      }).then(response => {
        saveAs(new Blob([response.data]), fileName);
      }).catch(error => {
        const message = parseServerError(error);
        this.setState({
          error: message
        });
      });
    };
  };

  _downloadAllFiles = _event => {
    const {sourceFiles} = this.state;
    const {type, runId} = this.props;
    if (Object.keys(sourceFiles).length > 0) {
      this.setState({
        isZipInProgress: true
      });
      backend({
        url: `api/v1/files/downloadAll/${runId}/${type}`,
        method: 'GET',
        responseType: 'blob'
      }).then(response => {
        this.setState({
          isZipInProgress: false
        });
        saveAs(new Blob([response.data]), `${type}-${runId}.zip`);
      }).catch(error => {
        const message = parseServerError(error);
        this.setState({
          isZipInProgress: false,
          error: message
        });
      });
    } else {
      this.setState({
        error: 'Error: No files are available to download'
      });
    }
  };

  componentDidMount() {
    setDbInfo(backend, this.props.dbInfo);
    this._fetchSourceFiles();
  }

  _handleAccordionItemChange = fileId => {
    const {sourceFiles} = this.state;
    const {files} = this.props;
    const fileInfo = files.find(file => file.file_id === fileId);
    const fileName = fileInfo.name || '';
    this.setState({
      accordionError: ''
    });
    if (sourceFiles[fileId] && !sourceFiles[fileId].data && sourceFiles[fileId].fileLength <= FILE_PREVIEW_LIMIT &&
        !imageExtensions.includes(getFileExtension(fileName))) {
      // Fetch contents of artifact/source file
      this.setState({
        isAccordionDataLoading: true
      });

      fetch(`${backend.defaults.baseURL}/api/v1/files/preview/${fileId}`)
        .then(response => {
          // Response body is a ReadableStream
          const reader = response.body.getReader();
          // Result is concatenated Array Buffer
          let result = null;
          const onStreamDone = () => {
            // Convert array buffer to string
            const resultText = new TextDecoder().decode(result);
            const updatedFileInfo = {...sourceFiles[fileId], data: resultText};
            this.setState(prevState => {
              return {
                sourceFiles: {...prevState.sourceFiles, [fileId]: updatedFileInfo},
                isAccordionDataLoading: false
              };
            });
          };

          reader.read().then(function processText({done, value}) {
            // Result objects contain two properties:
            // done  - true if the stream has already given you all its data.
            // value - some data. Always undefined when done is true.
            if (done) {
              onStreamDone();
              return;
            }

            // Concatenate array buffer
            result = concatArrayBuffers(result, value);

            // Read some more, and call this function again
            return reader.read().then(processText);
          });
        }).catch(error => {
          this.setState({
            isAccordionDataLoading: false,
            accordionError: parseServerError(error)
          });
        });
    }
  };

  _itemSizeEstimator = (_index, _cache) => {
    return MIN_ACCORDION_HEIGHT;
  };

  _renderAccordionItem = index => {
    const {files} = this.props;
    const {sourceFiles, accordionError, isAccordionDataLoading} = this.state;
    const file = files[index];
    return (
      <AccordionItem key={file.file_id} uuid={file.file_id} test-attr={'acc-item-' + index}>
        <AccordionItemTitle className='accordion__title accordion__title--animated'>
          <h5 className='u-position-relative'>
            {file.name}
            <div className='accordion__arrow' role='presentation'/>
          </h5>
          <div>{sourceFiles[file.file_id] && sourceFiles[file.file_id].filepath}</div>
        </AccordionItemTitle>
        <AccordionItemBody>
          <div className='clearfix'>
            <div className='pull-left upload-date'>Upload Date: {sourceFiles[file.file_id] && sourceFiles[file.file_id].uploadDate}</div>
            <div className='pull-right'>
              <Button test-attr={'down-btn-' + file.name} bsStyle='default' bsSize='xsmall' onClick={this._downloadFile(file.file_id, file.name)}>
                <i className='glyphicon glyphicon-download-alt'/> Download
              </Button>
            </div>
          </div>
          <FilePreview fileName={file.name} fileId={file.file_id} sourceFiles={sourceFiles}
            isLoading={isAccordionDataLoading} errorMessage={accordionError} dbInfo={this.props.dbInfo}/>
        </AccordionItemBody>
      </AccordionItem>
    );
  };

  render() {
    const {files, type} = this.props;
    const {isLoadingSourceFiles, isZipInProgress, error} = this.state;

    const errorAlert = error ? <Alert bsStyle='danger' test-attr='error-alert'>{error}</Alert> : '';
    const warningText = `Oops! There are no ${type} available for this run.`;
    const accordions = (
      <ProgressWrapper loading={isLoadingSourceFiles}>
        <div>
          <div className='download-all-wrapper clearfix'>
            <div className='pull-right'>
              <Button test-attr='down-all-btn' bsStyle='info' bsSize='small' disabled={isZipInProgress} onClick={this._downloadAllFiles}>
                {isZipInProgress ? <i className='glyphicon glyphicon-refresh glyphicon-refresh-animate'/> : <i className='glyphicon glyphicon-download-alt'/>}
              &nbsp;
                {isZipInProgress ? 'Archiving...' : 'Download All'}
              </Button>
            </div>
          </div>
          <Accordion accordion onChange={this._handleAccordionItemChange}>
            <ReactList
              itemRenderer={this._renderAccordionItem}
              length={files.length}
              type='variable'
              minSize={20}
              pageSize={10}
              threshold={300}
              itemSizeEstimator={this._itemSizeEstimator}
            />
          </Accordion>
        </div>
      </ProgressWrapper>
    );
    const content = files.length > 0 ? accordions : <Alert test-attr='warn-alert' bsStyle='warning'>{warningText}</Alert>;
    return (
      <div id='source-files-container'>
        {errorAlert}
        {content}
      </div>
    );
  }
}

export {SourceFilesView, FILE_PREVIEW_LIMIT};
