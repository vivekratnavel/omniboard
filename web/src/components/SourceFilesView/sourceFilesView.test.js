import React from 'react';
import mockAxios from 'jest-mock-axios';
import saveAs from 'file-saver';
import {parseServerError} from '../Helpers/utils';
import {SourceFilesView} from './sourceFilesView';

describe('SourceFilesView', () => {
  let wrapper = null;
  const type = 'artifacts';
  const runId = 1;
  const dbInfo = {key: 'default', name: 'test_db', path: '/test'};
  const files = [
    {
      name: '/src/hello_world.py',
      file_id: '5a1bd242ca100c1a210b0d3a'
    }
  ];
  const responseData = [
    {
      _id: '5a1bd242ca100c1a210b0d3a',
      uploadDate: '2017-11-27T08:52:18.562Z',
      md5: '6ef452da81287e70540ee3cf0bffd13e',
      length: 1646,
      chunkSize: 261120,
      filename: '/src/hello_world.py'
    },
    {
      _id: '5a1bd242ca100c1a210b0d3b',
      uploadDate: '2017-11-27T08:53:18.562Z',
      md5: '6ef452da81287e70540ee3cf9bffd13e',
      length: 0,
      chunkSize: 0,
      filename: '/src/hello_world_2.py'
    }
  ];

  beforeEach(() => {
    wrapper = mount(
      <SourceFilesView files={files} runId={runId} dbInfo={dbInfo} type={type}/>
    );
  });

  afterEach(() => {
    // Cleaning up the mess left behind the previous test
    mockAxios.reset();
    jest.clearAllMocks();
  });

  it('should render correctly', async () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state().isLoadingSourceFiles).toBeTruthy();
    mockAxios.mockResponse({status: 200, data: responseData});
    await tick();

    expect(wrapper.state().isLoadingSourceFiles).toBeFalsy();
    expect(wrapper.update()).toMatchSnapshot();
  });

  it('should handle error while fetching source files on mount', async () => {
    const errorResponse = {status: 404, message: 'unknown error'};
    mockAxios.mockError(errorResponse);

    expect(wrapper.state().error).toEqual(parseServerError(errorResponse));
  });

  describe('should download file', () => {
    it('when source exists', async () => {
      const data = '}qXresultqX\fHello world!qs.';
      mockAxios.mockResponse({status: 200, data: responseData});
      await tick();
      wrapper.update().find('[test-attr="acc-item-0"]').at(1).simulate('click');
      wrapper.update().find('[test-attr="down-btn-/src/hello_world.py"]').at(1).simulate('click');
      mockAxios.mockResponse({status: 200, data});
      await tick();

      expect(saveAs).toHaveBeenCalledWith(new Blob([data]), files[0].name);
    });

    it('display error when source does not exist', async () => {
      mockAxios.mockResponse({status: 200, data: []});
      await tick();
      wrapper.update().find('[test-attr="down-btn-/src/hello_world.py"]').at(1).simulate('click');
      const err = {status: 500, message: 'internal server error'};
      mockAxios.mockError(err);
      await tick();

      expect(saveAs).not.toHaveBeenCalled();
      expect(wrapper.state().error).toEqual(parseServerError(err));
    });
  });

  describe('should download all files', () => {
    it('when source exists', async () => {
      const data = '}qXresultqX\fHello world!qs.';
      mockAxios.mockResponse({status: 200, data: responseData});
      await tick();
      wrapper.update().find('[test-attr="down-all-btn"]').at(1).simulate('click');

      expect(wrapper.state().isZipInProgress).toBeTruthy();
      mockAxios.mockResponse({status: 200, data});
      await tick();

      expect(wrapper.state().isZipInProgress).toBeFalsy();
      expect(saveAs).toHaveBeenCalledWith(new Blob([data]), `${type}-${runId}.zip`);
    });

    it('display error when source does not exist', async () => {
      mockAxios.mockResponse({status: 200, data: []});
      await tick();
      wrapper.update().find('[test-attr="down-all-btn"]').at(1).simulate('click');
      await tick();

      expect(saveAs).not.toHaveBeenCalled();
      expect(wrapper.state().error).toEqual('Error: No files are available to download');
    });
  });

  describe('should handle accordion change', () => {
    it('and handle error', async () => {
      const fileId = responseData[0]._id;
      const error = {message: 'errorMessage'};
      mockAxios.mockResponse({status: 200, data: responseData});
      await tick();
      fetch.mockReject(error);
      wrapper.instance()._handleAccordionItemChange(fileId);

      expect(fetch.mock.calls).toHaveLength(1);
      expect(fetch.mock.calls[0][0]).toEqual(`/test/api/v1/files/preview/${fileId}`);

      await tick();

      expect(wrapper.update().state().isAccordionDataLoading).toBeFalsy();
      expect(wrapper.state().accordionError).toEqual(parseServerError(error));
    });
  });

  it('should show warning message when there are no files', async () => {
    wrapper = shallow(
      <SourceFilesView files={[]} runId={1} dbInfo={dbInfo} type='artifacts'/>
    );

    expect(wrapper.find('[test-attr="warn-alert"]')).toHaveLength(1);
  });

  it('should show error message when type is invalid', async () => {
    wrapper = shallow(
      <SourceFilesView files={files} runId={1} dbInfo={dbInfo} type='invalid'/>
    );

    expect(wrapper.find('[test-attr="error-alert"]')).toHaveLength(1);
  });
});
