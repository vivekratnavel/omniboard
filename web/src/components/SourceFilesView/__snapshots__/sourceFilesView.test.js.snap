// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceFilesView should render correctly 1`] = `
<SourceFilesView
  dbInfo={
    Object {
      "key": "default",
      "name": "test_db",
      "path": "/test",
    }
  }
  files={
    Array [
      Object {
        "file_id": "5a1bd242ca100c1a210b0d3a",
        "name": "/src/hello_world.py",
      },
    ]
  }
  runId={1}
  type="artifacts"
>
  <div
    id="source-files-container"
  >
    <ProgressWrapper
      loading={true}
    >
      <div
        className="table-display"
      >
        <div
          className="table-cell-display gray"
        >
          <i
            className="glyphicon glyphicon-refresh glyphicon-refresh-animate"
          />
           Loading...
        </div>
      </div>
    </ProgressWrapper>
  </div>
</SourceFilesView>
`;

exports[`SourceFilesView should render correctly 2`] = `
<SourceFilesView
  dbInfo={
    Object {
      "key": "default",
      "name": "test_db",
      "path": "/test",
    }
  }
  files={
    Array [
      Object {
        "file_id": "5a1bd242ca100c1a210b0d3a",
        "name": "/src/hello_world.py",
      },
    ]
  }
  runId={1}
  type="artifacts"
>
  <div
    id="source-files-container"
  >
    <ProgressWrapper
      loading={false}
    >
      <div>
        <div
          className="download-all-wrapper clearfix"
        >
          <div
            className="pull-right"
          >
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsSize="small"
              bsStyle="info"
              disabled={false}
              onClick={[Function]}
              test-attr="down-all-btn"
            >
              <button
                className="btn btn-sm btn-info"
                disabled={false}
                onClick={[Function]}
                test-attr="down-all-btn"
                type="button"
              >
                <i
                  className="glyphicon glyphicon-download-alt"
                />
                Â 
                Download All
              </button>
            </Button>
          </div>
        </div>
        <AccordionWrapper
          accordion={true}
          className="accordion"
          onChange={[Function]}
        >
          <Provider
            accordion={true}
            onChange={[Function]}
          >
            <Consumer>
              <Accordion
                accordion={true}
                className="accordion"
              >
                <div
                  className="accordion"
                  role="tablist"
                >
                  <ReactList
                    axis="y"
                    itemRenderer={[Function]}
                    itemSizeEstimator={[Function]}
                    itemsRenderer={[Function]}
                    length={1}
                    minSize={20}
                    pageSize={10}
                    scrollParentGetter={[Function]}
                    scrollParentViewportSizeGetter={[Function]}
                    threshold={300}
                    type="variable"
                    useStaticSize={false}
                    useTranslate3d={false}
                  >
                    <div
                      style={
                        Object {
                          "height": 51,
                          "position": "relative",
                        }
                      }
                    >
                      <div
                        style={
                          Object {
                            "WebkitTransform": "translate(0px, 0px)",
                            "msTransform": "translate(0px, 0px)",
                            "transform": "translate(0px, 0px)",
                          }
                        }
                      >
                        <div>
                          <FromRenderProps
                            key="5a1bd242ca100c1a210b0d3a"
                            test-attr="acc-item-0"
                            uuid="5a1bd242ca100c1a210b0d3a"
                          >
                            <Consumer>
                              <AccordionItemWrapper
                                accordionStore={
                                  Object {
                                    "accordion": true,
                                    "addItem": [Function],
                                    "items": Array [
                                      Object {
                                        "disabled": false,
                                        "expanded": false,
                                        "uuid": "5a1bd242ca100c1a210b0d3a",
                                      },
                                    ],
                                    "removeItem": [Function],
                                    "setExpanded": [Function],
                                  }
                                }
                                className="accordion__item"
                                disabled={false}
                                expanded={false}
                                hideBodyClassName=""
                                test-attr="acc-item-0"
                                uuid="5a1bd242ca100c1a210b0d3a"
                              >
                                <Provider
                                  uuid="5a1bd242ca100c1a210b0d3a"
                                >
                                  <AccordionItem
                                    accordionStore={
                                      Object {
                                        "accordion": true,
                                        "addItem": [Function],
                                        "items": Array [
                                          Object {
                                            "disabled": false,
                                            "expanded": false,
                                            "uuid": "5a1bd242ca100c1a210b0d3a",
                                          },
                                        ],
                                        "removeItem": [Function],
                                        "setExpanded": [Function],
                                      }
                                    }
                                    className="accordion__item"
                                    disabled={false}
                                    expanded={false}
                                    hideBodyClassName=""
                                    test-attr="acc-item-0"
                                    uuid="5a1bd242ca100c1a210b0d3a"
                                  >
                                    <div
                                      className="accordion__item"
                                      test-attr="acc-item-0"
                                    >
                                      <FromRenderProps
                                        className="accordion__title accordion__title--animated"
                                      >
                                        <Consumer>
                                          <FromRenderProps
                                            accordionStore={
                                              Object {
                                                "accordion": true,
                                                "addItem": [Function],
                                                "items": Array [
                                                  Object {
                                                    "disabled": false,
                                                    "expanded": false,
                                                    "uuid": "5a1bd242ca100c1a210b0d3a",
                                                  },
                                                ],
                                                "removeItem": [Function],
                                                "setExpanded": [Function],
                                              }
                                            }
                                            className="accordion__title accordion__title--animated"
                                          >
                                            <Consumer>
                                              <AccordionItemTitleWrapper
                                                accordionStore={
                                                  Object {
                                                    "accordion": true,
                                                    "addItem": [Function],
                                                    "items": Array [
                                                      Object {
                                                        "disabled": false,
                                                        "expanded": false,
                                                        "uuid": "5a1bd242ca100c1a210b0d3a",
                                                      },
                                                    ],
                                                    "removeItem": [Function],
                                                    "setExpanded": [Function],
                                                  }
                                                }
                                                className="accordion__title accordion__title--animated"
                                                hideBodyClassName=""
                                                itemStore={
                                                  Object {
                                                    "uuid": "5a1bd242ca100c1a210b0d3a",
                                                  }
                                                }
                                              >
                                                <AccordionItemTitle
                                                  accordion={true}
                                                  className="accordion__title accordion__title--animated"
                                                  disabled={false}
                                                  expanded={false}
                                                  hideBodyClassName=""
                                                  setExpanded={[Function]}
                                                  uuid="5a1bd242ca100c1a210b0d3a"
                                                >
                                                  <div
                                                    aria-controls="accordion__body-5a1bd242ca100c1a210b0d3a"
                                                    aria-selected={false}
                                                    className="accordion__title accordion__title--animated"
                                                    disabled={false}
                                                    id="accordion__title-5a1bd242ca100c1a210b0d3a"
                                                    onClick={[Function]}
                                                    onKeyPress={[Function]}
                                                    role="tab"
                                                    tabIndex="0"
                                                  >
                                                    <h5
                                                      className="u-position-relative"
                                                    >
                                                      /src/hello_world.py
                                                      <div
                                                        className="accordion__arrow"
                                                        role="presentation"
                                                      />
                                                    </h5>
                                                    <div>
                                                      /src/hello_world.py
                                                    </div>
                                                  </div>
                                                </AccordionItemTitle>
                                              </AccordionItemTitleWrapper>
                                            </Consumer>
                                          </FromRenderProps>
                                        </Consumer>
                                      </FromRenderProps>
                                      <FromRenderProps>
                                        <Consumer>
                                          <FromRenderProps
                                            accordionStore={
                                              Object {
                                                "accordion": true,
                                                "addItem": [Function],
                                                "items": Array [
                                                  Object {
                                                    "disabled": false,
                                                    "expanded": false,
                                                    "uuid": "5a1bd242ca100c1a210b0d3a",
                                                  },
                                                ],
                                                "removeItem": [Function],
                                                "setExpanded": [Function],
                                              }
                                            }
                                          >
                                            <Consumer>
                                              <AccordionItemBodyWrapper
                                                accordionStore={
                                                  Object {
                                                    "accordion": true,
                                                    "addItem": [Function],
                                                    "items": Array [
                                                      Object {
                                                        "disabled": false,
                                                        "expanded": false,
                                                        "uuid": "5a1bd242ca100c1a210b0d3a",
                                                      },
                                                    ],
                                                    "removeItem": [Function],
                                                    "setExpanded": [Function],
                                                  }
                                                }
                                                className="accordion__body"
                                                hideBodyClassName="accordion__body--hidden"
                                                itemStore={
                                                  Object {
                                                    "uuid": "5a1bd242ca100c1a210b0d3a",
                                                  }
                                                }
                                              >
                                                <AccordionItemBody
                                                  accordion={true}
                                                  className="accordion__body"
                                                  disabled={false}
                                                  expanded={false}
                                                  hideBodyClassName="accordion__body--hidden"
                                                  uuid="5a1bd242ca100c1a210b0d3a"
                                                >
                                                  <div
                                                    aria-hidden={true}
                                                    aria-labelledby="accordion__title-5a1bd242ca100c1a210b0d3a"
                                                    className="accordion__body accordion__body--hidden"
                                                    id="accordion__body-5a1bd242ca100c1a210b0d3a"
                                                    role="tabpanel"
                                                  >
                                                    <div
                                                      className="clearfix"
                                                    >
                                                      <div
                                                        className="pull-left upload-date"
                                                      >
                                                        Upload Date: 
                                                        2017-11-27T08:52:18.562Z
                                                      </div>
                                                      <div
                                                        className="pull-right"
                                                      >
                                                        <Button
                                                          active={false}
                                                          block={false}
                                                          bsClass="btn"
                                                          bsSize="xsmall"
                                                          bsStyle="default"
                                                          disabled={false}
                                                          onClick={[Function]}
                                                          test-attr="down-btn-/src/hello_world.py"
                                                        >
                                                          <button
                                                            className="btn btn-xs btn-default"
                                                            disabled={false}
                                                            onClick={[Function]}
                                                            test-attr="down-btn-/src/hello_world.py"
                                                            type="button"
                                                          >
                                                            <i
                                                              className="glyphicon glyphicon-download-alt"
                                                            />
                                                             Download
                                                          </button>
                                                        </Button>
                                                      </div>
                                                    </div>
                                                    <FilePreview
                                                      dbInfo={
                                                        Object {
                                                          "key": "default",
                                                          "name": "test_db",
                                                          "path": "/test",
                                                        }
                                                      }
                                                      errorMessage=""
                                                      fileId="5a1bd242ca100c1a210b0d3a"
                                                      fileName="/src/hello_world.py"
                                                      isLoading={false}
                                                      sourceFiles={
                                                        Object {
                                                          "5a1bd242ca100c1a210b0d3a": Object {
                                                            "chunkSize": 261120,
                                                            "fileLength": 1646,
                                                            "filepath": "/src/hello_world.py",
                                                            "id": "5a1bd242ca100c1a210b0d3a",
                                                            "uploadDate": "2017-11-27T08:52:18.562Z",
                                                          },
                                                          "5a1bd242ca100c1a210b0d3b": Object {
                                                            "chunkSize": 0,
                                                            "fileLength": 0,
                                                            "filepath": "/src/hello_world_2.py",
                                                            "id": "5a1bd242ca100c1a210b0d3b",
                                                            "uploadDate": "2017-11-27T08:53:18.562Z",
                                                          },
                                                        }
                                                      }
                                                    >
                                                      <div>
                                                        <ProgressWrapper
                                                          loading={false}
                                                        />
                                                      </div>
                                                    </FilePreview>
                                                  </div>
                                                </AccordionItemBody>
                                              </AccordionItemBodyWrapper>
                                            </Consumer>
                                          </FromRenderProps>
                                        </Consumer>
                                      </FromRenderProps>
                                    </div>
                                  </AccordionItem>
                                </Provider>
                              </AccordionItemWrapper>
                            </Consumer>
                          </FromRenderProps>
                        </div>
                      </div>
                    </div>
                  </ReactList>
                </div>
              </Accordion>
            </Consumer>
          </Provider>
        </AccordionWrapper>
      </div>
    </ProgressWrapper>
  </div>
</SourceFilesView>
`;
